src/main/java/net/minecraftforge/common/world/RegistryAccessExtension.java

```diff
@@ -0,0 +1,123 @@
+ * provided at {@link #createRegistryKey(String, String)} and {@link #createRegistryKey(ResourceLocation)}.
+ * <p>
+ * Default or 'built-in' registry entries can be registered via a forge DeferredRegister or IForgeRegistry
+ * of the same registry ResourceKey. These elements will populate the resulting RegistryAccess before datapacks
+ * are loaded and may be overridden by data-definitions. A static helper method for creating a DeferredRegisters
+ * for a RegistryAccessExtension is provided at {@link #createRegister(Class, ResourceKey)}.
+ *
+ * @param <T> The type of world-gen component.
+ */
+public class RegistryAccessExtension<T extends IForgeRegistryEntry<T>> extends ForgeRegistryEntry<RegistryAccessExtension<?>>
+{
+    private final ResourceKey<Registry<T>> registryKey;
+    private final Codec<T> directCodec;
+    private final Lifecycle defaultElementLifecycle;
+
+    public RegistryAccessExtension(ResourceKey<Registry<T>> registryKey, Codec<T> directCodec)
+    {
+        this(registryKey, directCodec, Lifecycle.experimental());
+    }
+
+    public RegistryAccessExtension(ResourceKey<Registry<T>> registryKey, Codec<T> directCodec, Lifecycle defaultElementLifecycle)
+    {
+        validateRegistryKey(registryKey);
+        setRegistryName(registryKey.location());
+        this.registryKey = registryKey;
+        this.directCodec = directCodec;
+        this.defaultElementLifecycle = defaultElementLifecycle;
+    }
+
+    public ResourceKey<? extends Registry<T>> getRegistryKey()
+    {
+        return registryKey;
+    }
+
+    public Codec<T> getDirectCodec()
+    {
+        return directCodec;
+    }
+
+    public Lifecycle getDefaultElementLifecycle()
+    {
+        return defaultElementLifecycle;
+    }
+
+    public static <T> ResourceKey<Registry<T>> createRegistryKey(String namespace, String path)
+    {
+        var prefix = getPathPrefix(namespace);
+        if (!path.startsWith(prefix))
+        {
+            path = prefix + path;
+        }
+
+        return ResourceKey.createRegistryKey(new ResourceLocation(namespace, path));
+    }
+
+    public static <T> ResourceKey<Registry<T>> createRegistryKey(ResourceLocation location)
+    {
+        return createRegistryKey(location.getNamespace(), location.getPath());
+    }
+
+    public static <T extends IForgeRegistryEntry<T>> DeferredRegister<T> createRegister(Class<T> type, ResourceKey<Registry<T>> key)
+    {
+        validateRegistryKey(key);
+        var register = DeferredRegister.create(type, key.location().getNamespace());
+        register.makeRegistry(key.location().getPath(), () -> new RegistryBuilder<T>().disableSaving().disableSync());
+        return register;
+    }
+
+    private static String getPathPrefix(String namespace)
+    {
+        return namespace + '/';
+    }
+
+    private static void validateRegistryKey(ResourceKey<? extends Registry<?>> registryKey)
```

Comment:
This is just a comment making sure there are no complaints on the current standard before moving forward since there was a discussion about it above.